### PR TITLE
rgw: lua, add support for lua packages to the lua manager

### PR DESCRIPTION
**Pull request by kalebskeithley**
Refactor lua manager, adding new {add,remove}_package(), and
list_packages() methods; and changing get(), put(), del() methods
to {get,put,del}_script().  Renaming class from *LuaScriptManager
to *LuaManager to be consistent with dual *script and *package
functionality.

Note: I considered keeping the existing *LuaScriptManager classes
as is and adding *LuaPackageManager classes, but that seemed/seems
like overkill to me.

This PR supercedes https://github.com/ceph/ceph/pull/46980

Signed-off-by: Kaleb S. KEITHLEY <kkeithle@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by dang — APPROVED** [review]

File: src/rgw/rgw_lua.cc [hunk 1]
```diff
@@ -5,11 +5,8 @@
 #include "rgw_lua_utils.h"
 #include "rgw_sal_rados.h"
 #include "rgw_lua.h"
-#ifdef WITH_RADOSGW_LUA_PACKAGES
```
Comment (yuvalif) [inline src/rgw/rgw_lua.cc:8]: now that the luarock code is in this file, we need the WITH_RADOSGW_LUA_PACKAGES macros

File: src/rgw/rgw_sal_store.h [hunk 2]
```diff
@@ -404,9 +404,15 @@ class StoreZone : public Zone {
     virtual ~StoreZone() = default;
 };
 
-class StoreLuaScriptManager : public LuaScriptManager {
+class StoreLuaManager : public LuaManager {
 public:
-  virtual ~StoreLuaScriptManager() = default;
+  virtual ~StoreLuaManager() = default;
+
+  virtual int add_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name, bool allow_compilation) { return -ENOENT; }
```
Comment (yuvalif) [inline src/rgw/rgw_sal_store.h:411]: why do we have stub implementations to the package related functions while the script related functions are pure virtual?

File: src/rgw/rgw_lua.cc (hunk 1 above)
Comment (kalebskeithley) [reply src/rgw/rgw_lua.cc:8]: okay

File: src/rgw/rgw_sal_store.h (hunk 2 above)
Comment (kalebskeithley) [reply src/rgw/rgw_sal_store.h:418]: do you want them all stubs, or all pure virtual here?

File: src/rgw/rgw_sal_rados.cc [hunk 3]
```diff
@@ -3056,6 +3058,119 @@ int RadosLuaScriptManager::del(const DoutPrefixProvider* dpp, optional_yield y,
...
+    return -EINVAL;
+  }
+
+  //replace previous versions of the package
+  const std::string package_name_no_version = package_name.substr(0, package_name.find(" "));
+  ret = remove_package(dpp, y, package_name_no_version);
+  if (ret < 0) {
+    return ret;
+  }
+
```
Comment (cbodley) [inline src/rgw/rgw_sal_rados.cc:3101]: copied from https://github.com/ceph/ceph/pull/46980#discussion_r921344296:

> please leave all of this code where it was (above the zipper line)

File: src/rgw/rgw_sal_rados.h [hunk 4]
```diff
@@ -869,17 +869,22 @@ class RadosMultipartWriter : public StoreWriter {
...
-  virtual int get(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, std::string& script) override;
-  virtual int put(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, const std::string& script) override;
-  virtual int del(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key) override;
+  RadosLuaManager(RadosStore* _s);
+  virtual ~RadosLuaManager() = default;
+
+  virtual int get_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, std::string& script) override;
+  virtual int put_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, const std::string& script) override;
+  virtual int del_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key) override;
+#ifdef WITH_RADOSGW_LUA_PACKAGES
```
Comment (cbodley) [inline src/rgw/rgw_sal_rados.h:883]: please drop the #ifdefs in rgw_sal*

rgw_lua.cc should be the only place we're shelling out to the `luarocks` command

File: src/rgw/rgw_sal_rados.cc (hunk 3 above)
Comment (kalebskeithley) [reply src/rgw/rgw_sal_rados.cc]: okay

File: src/rgw/rgw_sal_rados.h (hunk 4 above)
Comment (kalebskeithley) [reply src/rgw/rgw_sal_rados.h]: done

File: src/rgw/rgw_sal.h [hunk 5]
```diff
@@ -1500,20 +1501,26 @@ class Zone {
...
+  virtual int put_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, const std::string& script) = 0;
   /** Delete a script named with the given key from the backing store */
-  virtual int del(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key) = 0;
+  virtual int del_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key) = 0;
+  /** Add a lua package */
+  virtual int add_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name) = 0;
+  /** Remove a lua package */
+  virtual int remove_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name) = 0;
+  /** List lua packages */
+  virtual int list_packages(const DoutPrefixProvider* dpp, optional_yield y, rgw::lua::packages_t& packages) = 0;
```
Comment (cbodley) [inline src/rgw/rgw_sal.h:1523]: this refers to `rgw::lua::packages_t` which is declared inside the ifdef in rgw_lua.h:
```c++
#ifdef WITH_RADOSGW_LUA_PACKAGES
#include <set>

using packages_t = std::set<std::string>;
```
we'll need to move that out of the ifdef for use in sal

**Review by cbodley — COMMENTED** [review]
:+1:

File: src/rgw/rgw_sal.h (hunk 5 above)
Comment (kalebskeithley) [reply src/rgw/rgw_sal.h:1523]: indeed

File: src/rgw/rgw_sal_store.h [hunk 6]
```diff
@@ -404,9 +404,22 @@ class StoreZone : public Zone {
...
 };
 
-class StoreLuaScriptManager : public LuaScriptManager {
+class StoreLuaManager : public LuaManager {
 public:
-  virtual ~StoreLuaScriptManager() = default;
+  virtual ~StoreLuaManager() = default;
+
+  /** Get a script named with the given key from the backing store */
+  virtual int get_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, std::string& script) { return -ENOENT; }
```
Comment (yuvalif) [inline src/rgw/rgw_sal_store.h:412]: if these stubs are implemented here, we can probably avoid them in motr dbstore?

**Review by yuvalif — APPROVED** [review]
please remove stubs from as_motr and sal_dbstore as they are not implemented in sal_store

File: src/rgw/rgw_sal_store.h (hunk 6 above)
Comment (kalebskeithley) [reply src/rgw/rgw_sal_store.h:412]: okay. My C++ is still pretty rusty. ;-)

**Pull request closed without merging**